src/evict/evict_page.c

```diff
@@ -675,9 +676,37 @@ __evict_review(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t evict_flags, bool
             LF_SET(WT_REC_SCRUB);
     }
 
+    /* Acquire a snapshot if coming through eviction thread route. */
+    if (F_ISSET(session, WT_SESSION_INTERNAL) && !F_ISSET(session->txn, WT_TXN_RUNNING) &&
+      !F_ISSET(conn,
+        WT_CONN_CLOSING | WT_CONN_RECOVERING | WT_CONN_IN_MEMORY | WT_CONN_CLOSING_TIMESTAMP) &&
+      !WT_IS_HS(S2BT(session))) {
+        /*
+         * We don't have a running transaction with eviction threads. Starting a new transaction
+         * should implicitly take a snapshot as well.
+         */
+        WT_RET(__wt_txn_begin(session, txn_cfg));
+        /* Eviction threads do not need to pin anything. */
+        WT_SESSION_TXN_SHARED(session)->pinned_id = WT_TXN_NONE;
+        txn_started = true;
+    } else {
+        /*
+         * Only set this flag for application threads or whenever we are unable to start a
+         * transaction.
+         */
+        LF_SET(WT_REC_VISIBLE_ALL);
+    }
+    WT_ASSERT(session, !txn_started || F_ISSET(session->txn, WT_TXN_HAS_SNAPSHOT));
```

Comment:
If an application starts a transaction in a read committed mode, the session transaction set WT_TXN_RUNNING, but I am not sure whether it acquires the snapshot before it is used for evicting the pages from the cache. In case if it is then this assert may get hit.

Reply:
This change is not covering application threads that have entered the eviction code. For now we are only dealing with dedicated eviction threads.